### ISSUE TITLE
edge case checks for data.table

### ISSUE THREAD

**Issue by jan-glx** [Rdatatable/data.table#4705]
`data.table` handles zero row data.tables. But to know which columns the result should have, `data.table` evaluates `j` for zero length data.tables. This behavior might be unexpected when grouping since j is evaluated for a zero row "group" only if there is no group with more than zero rows because one might have thought of using `data.table` with `by` as a replacement of a `for` loop.
Unfortunately, `R` is filled with edge case bugs, so evaluating `j` with zero length variables will often cause errors. These are hard to detect if they only surface when a data.table has zero rows. I would therefore propose to always evaluate `j` for a zero row version of the data.table as well and warn/message the user if the column types of the output differs (and point to an parameter to turn off this check).

For example,
```
data.table(k=1, n=2)[, .(pval = prop.test(k, n)$p.value)]
#    pval
# 1:    1
# Warning message:
# In prop.test(k, n) : Chi-squared approximation may be incorrect
```
would detect this error:
```
data.table(k=1, n=2)[FALSE, ][, .(pval = prop.test(k, n)$p.value)]
# Error in prop.test(k, n) : not enough data
```
and warn with something like:
```
# Warning message:
# In `[`( , j= .(pval = prop.test(k, n)$p.value)) : j throws an error if DT has zero rows instead of `.(pval = numeric(0))`, consider fixing the expression in `j` to handle this edge case or hide this warning using `.check_zero_row = FALSE`! See #XXX for more information.
```
such the the programmer could implement such a workaround:
```
data.table(k=1, n=2)[FALSE, ][, .(pval = if(.N<1) 1 else prop.test(k, n)$p.value)]
#    pval
# 1:    1
```
or explicitly ignore it like e.g.:
```
data.table(k=1, n=2)[, .(pval = prop.test(k, n)$p.value), .check_zero_row = FALSE]
```

**Comment by jan-glx** [comment]
An alternative solution would allow to specify the return type of `j` in a separate argument like with `FUN.VALUE` of `vapply` to use instead of evaluating `j` for zero row data.tables.

**Comment by ben-schwen** [comment]
I don't think that this is unexpected behavior or an error since `base` gives you the same result.

``` r
df = data.frame(k=numeric(0), n=numeric(0))
pval = prop.test(df$k, df$n)$p.value
#> Error in prop.test(df$k, df$n): not enough data
```

Therefore closing, but feel free to reopen if I have overlooked smth.
